### PR TITLE
Add explicit support for length attribute

### LINKED ISSUE #202: continue to deal the result as a array after filter
## Motivation
I have a json like this:
`
{
  "contactMedium": [
    {
      "@type": "EmailContactMedium",
      "preferred": true,
      "contactType": "other",
      "validFor": {
        "startDateTime": "2018-10-22T08:31:52.028Z"
      },
      "emailAddress": "tom@email.com"
    },
    {
      "@type": "GeographicAddressContactMedium",
      "preferred": true,
      "contactType": "other",
      "validFor": {
        "startDateTime": "2018-10-22T08:31:52.028Z"
      },
    }
  ]
}
`
I want to get the number of item with preferred==true
const result = JSONPath({ path: '$.contactMedium[?(@.preferred==true)].length', json: data })
## Current behavior
got []
## Desired behavior
got 2
## Alternatives considered
currently, i can got the result expected by this way.
const result = JSONPath({ path: '$.contactMedium[?(@.preferred==true)]', json: data })
console.log(result)

const result2 = JSONPath({ path: '$.length', json: result })
console.log(result2)

### PR DESCRIPTION
**Pull request by gerad**
## PR description

The length attribute works by happy accident today. This adds initial explicit support for it and also gets it work with filter conditions.

This is not complete support. It won't (for instance) work correctly within array filter conditions. See the test changes for use cases it does add support for.

Note that the ietf standardization for length() and count() functions is done differently.

See: https://ietf-wg-jsonpath.github.io/draft-ietf-jsonpath-base/draft-ietf-jsonpath-base.html#name-length-function-extension

Fixes #202

## Checklist

- [x] - Added tests
- [x] - Ran `npm test`, ensuring linting passes
- [x] - Adjust README documentation if relevant

### CODE REVIEW

File: dist/index-browser-esm.js [hunk 1]
```diff
@@ -1586,10 +1586,25 @@ JSONPath.prototype.evaluate = function (expr, json, callback, otherTypeCallback)
...
   this._hasParentSelector = null;
-  const result = this._trace(exprList, json, ['$'], currParent, currParentProperty, callback).filter(function (ea) {
+  let result = this._trace(exprList, json, ['$'], currParent, currParentProperty, callback).filter(function (ea) {
     return ea && !ea.isParentSelector;
   });
+  if (wantsLength) {
+    const res0 = result[0];
+    result = [{
+      path: push(res0.path, 'length'),
+      value: result.length,
```
Comment (gerad) [inline dist/index-browser-esm.js:1602]: do we need to flatten here?

File: src/jsonpath.js [hunk 2]
```diff
@@ -260,15 +260,31 @@ JSONPath.prototype.evaluate = function (
     if (exprList[0] === '$' && exprList.length > 1) {
         exprList.shift();
     }
+    let wantsLength = false;
+    if (exprList.at(-1) === 'length') {
```
Comment (branets-dasha) [inline src/jsonpath.js:264]: I think treating `.length` as length function is a bad idea, it will conflict with `length` property, if the object has it.

**Comment by brettz9** [comment]
I think we really need to freeze development which does not align with the IETF spec (unless that spec is truly deficient in some way and in such a case, follow the jsonpath comparison project as possible for the way forward).

**Comment by gerad** [comment]
Thanks for the fast feedback both! Good point on objects with a length property. 

Will revisit.

**Pull request closed without merging**